### PR TITLE
Add Polynomial parameters

### LINKED ISSUE #334: y=mx+c parameter
I've seen quite a few inflows that are based on regression from another inflow. It's ugly/slow to create these using the existing AggregatedParameter objects. It would be nice to have a parameter that could calculate y=mx+c where y is the new flow and x is the original flow.

In theory we could implement something more complex, that took an arbitrary equation. I don't think this is needed right now though.

### PR DESCRIPTION
**Pull request by jetuk**
Two generic polynomial parameters.

The 1D parameter can be related to either a node (for flow), storage (for volume) or another parameter. The 2D parameter is a bit more specialised and uses a storage and parameter for the 2 dimensions. I couldn't think of a clean way to make the choice of the dimensions general while also being efficient.

Fixes #334

### CODE REVIEW

File: pywr/parameters/_polynomial.pyx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        self._parameter = kwargs.pop('parameter', None)
+        self.use_proportional_volume = kwargs.pop('use_proportional_volume', False)
+        # Check only one of the above is given
+        arg_check = [
+            self._other_node is not None,
+            self._storage_node is not None,
+            self._parameter is not None,
+        ]
+        # Check we haven't been given an ambiguous number of objects
+        if arg_check.count(True) > 1:
```
Comment (snorfalorpagus) [inline pywr/parameters/_polynomial.pyx]: Should this be == 1? What if the user doesn't pass anything?

File: pywr/parameters/_polynomial.pyx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    parameter : iterable of Parameter objects or single Parameter
+        An `Parameter` the value of which is used to evaluate the polynomial.
+    use_proportional_volume : bool
+        An optional boolean only used with a `Storage` node to switch between using absolute
+         or proportional volume when evaluating the polynomial.
+    """
+    def __init__(self, coefficients, storage_node, parameter, *args, **kwargs):
+        self.coefficients = np.array(coefficients, dtype=np.float64)
+        self._storage_node = storage_node
+        self._parameter = parameter
```
Comment (snorfalorpagus) [inline pywr/parameters/_polynomial.pyx]: The parameter should be added to this parameter's children. `self.children.add(parameter)`

File: pywr/parameters/_polynomial.pyx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        An optional `Parameter` the value of which is used to evaluate the polynomial.
+    use_proportional_volume : bool
+        An optional boolean only used with a `Storage` node to switch between using absolute
+         or proportional volume when evaluating the polynomial.
+    """
+    def __init__(self, coefficients, *args, **kwargs):
+        self.coefficients = np.array(coefficients, dtype=np.float64)
+        self._other_node = kwargs.pop('node', None)
+        self._storage_node = kwargs.pop('storage_node', None)
+        self._parameter = kwargs.pop('parameter', None)
```
Comment (snorfalorpagus) [inline pywr/parameters/_polynomial.pyx]: The parameter should be added to this parameter's children. `self.children.add(parameter)`

**Comment by snorfalorpagus** [comment]
This looks OK, with the exception of the comments I've added. It should work nicely with #362. If you make those changes I'll give it another read through before merging when I'm more awake!

File: pywr/parameters/_polynomial.pyx (hunk 1 above)
Comment (jetuk) [inline pywr/parameters/_polynomial.pyx]: Passing nothing is valid.  It uses `self._node.flow` instead. See line 62.

File: pywr/parameters/_polynomial.pyx (hunk 2 above)
Comment (jetuk) [inline pywr/parameters/_polynomial.pyx]: True, and applies to the 1D class too.

**Comment by jetuk** [comment]
I've made the changes to the above, except the one I don't agree with.

I've also added the parameters discussed in #362 to this PR.

File: pywr/parameters/_polynomial.pyx (hunk 3 above)
Comment (snorfalorpagus) [inline pywr/parameters/_polynomial.pyx]: This change has been addressed (on a different line).

File: tests/test_parameters.py [hunk 4]
```diff
@@ -686,3 +688,162 @@ def test_invalid_parameter_values():
...
+
+    def test_1st_order_with_parameter(self, model):
+        """ Test 1st order with a `Parameter` """
+        p1 = Polynomial1DParameter([0.5, np.pi], parameter=ConstantParameter(2.0))
+        si = ScenarioIndex(0, np.array([0], dtype=np.int32))
+        ts = model.timestepper.current
+        np.testing.assert_allclose(p1.value(ts, si), 0.5 + np.pi*2.0)
+
+    def test_2nd_order_with_parameter(self, model):
+        """ Test 2nd order with a `Parameter` """
```
Comment (snorfalorpagus) [inline tests/test_parameters.py]: This test could be made more clear if you used a variable when defining the constant parameter (e.g. `x = 2.0` `p1 = ConstantParameter(x)` ... `0.5 + np.pi*x + 3.0*x**2`). Also, don't redefine p1.

**Comment by snorfalorpagus** [comment]
I've made one more comment. I can't see anything wrong with the code, it's just a style thing.

File: tests/test_parameters.py (hunk 4 above)
Comment (jetuk) [inline tests/test_parameters.py]: Addressed in the latest commit.

**Pull request merged**